DB-9615 Add file logging to SQL shell.

hatyo opened the pull request:
- Add commands logto, logstop, and logclear which give the user
  the possibility to log SQL statements to a local file.
- logto requires quoting the file path, i.e. logto '/tmp/log.txt'

This is still WIP, but I still wanted to take your input on this change, specially the way I extended the ij parser rules for the new logging commands.

hatyo commented:
Hi Oleg, sure it makes sense to reuse logging, could you maybe provide me with a couple of pointers about Derby's client-side logging that you mentioned so I could have a look?

hatyo commented:
thanks for the explanation, I am not sure the JIRA ticket requirement aligns with normal logging, I thought it is e.g. similar to GDB's [file logging](https://sourceware.org/gdb/onlinedocs/gdb/Logging-Output.html). i.e. we want only to allow record user interactions with interactive shell and save it to a file.
To me this is different from normal (log4j) logging although we might argue that we could also leverage (log4j) logging to do the same, i.e. by having an explicit log-level for SQL statements and their output written somewhere so it can be picked up later using some log-aggregation framework.
What do you think?

dgomezferro commented:
@hatyo I agree this is different from "execution tracing" logging, and I think a separate feature independent of log4j makes sense (one less dependency too for such a simple feature). Other DBs' shells have similar features:
- postgresql: \out command
- oracle: SPOOL command https://docs.oracle.com/cd/B19306_01/server.102/b14357/ch12043.htm

hatyo commented:
I updated the PR and changed the commands as the following:
* `logto <filename>` --> `spool <filename>`
* `logstop` --> `spool stop`
* `logclear` --> `spool clear`

hatyo commented:
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

spliceci commented:
TEST SUCCEEDED +1

hatyo commented on db-tools-ij/src/main/grammar/ij.jj:
```diff
@@ -3374,6 +3378,7 @@ HelpStatement() :
 		while (st.hasMoreTokens()) {
 		    v.addElement(st.nextToken());
 		}
+		v.addElement("Something");
```
I was experimenting with something and this slipped through, thanks for finding it out.

hatyo commented on db-tools-ij/src/main/grammar/ij.jj:
```diff
@@ -4043,7 +4061,42 @@ ijResult CP_DisconnectStatement() throws SQLException
 
 }
 
+ijResult LogToStatement() throws IOException
```
That makes sense, thanks.

arnaud-splice commented on db-drda/src/main/java/com/splicemachine/db/impl/tools/sysinfo/Main.java:
```diff
@@ -1027,52 +1014,42 @@ private static ZipInfoProperties checkDirectory(String dirname)
-
             if (thisEntry == null)
             {
+                zf.close();
```
You don't need this zf.close anymore

dgomezferro commented on db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java:
```diff
@@ -845,4 +820,26 @@ int getCurrentRowNumber(ResultSet rs)
+
+	public void stopSpooling() {
+		doSpool = false;
+		this.out.flush();
+		this.out = new LocalizedOutput(System.out);
```
We should close the FileOutputStream, relying on finalizers is very finicky

hatyo replied:
ok

dgomezferro commented on db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java:
```diff
@@ -320,6 +295,11 @@ private int runScriptGuts() {
 				out.flush();
 				command = commandGrabber[currCE].nextStatement();
 
+				if(doSpool) {
```
Does this echo the commands to the terminal too? Is that what we want?

hatyo replied:
From the JIRA ticket:
> they asked for a function that would allow them to log sqlshell commands AND results to a file on demand

Let's discuss improvements in a different JIRA maybe?

dgomezferro replied:
I mean this code seems to be echoing the typed command back to the terminal, so the users would see what they typed twice (once when they type it, once when it's echoed) and then the result. The spooled file would have the command once and the result once, which is correct. I might be wrong and we might not be echoing the command to the terminal again, but I wanted to make sure whether that's the case.

hatyo replied:
Hi Daniel,
sorry I misread you comment, you're right, we're writing the command twice to tty, I will adapt the change accordingly.

dgomezferro replied:
No worries! Change looks good, thanks!

dgomezferro commented:
This is probably not the place for the discussion, but it seems one use case for Oracle's SPOOL is to create SQL scripts dynamically, which we won't be able to do because we are writing the original SQL query too.

hatyo commented:
> This is probably not the place for the discussion, but it seems one use case for Oracle's SPOOL is to create SQL scripts dynamically, which we won't be able to do because we are writing the original SQL query too.

Yeah it makes sense to have more control about what's reported. I think Michael had some ideas and maybe we could address them in a another improvement ticket.
